enterprise/server/githubapp/githubapp.go

```diff
@@ -586,6 +592,101 @@ func (a *GitHubApp) GetAccessibleGitHubRepos(ctx context.Context, req *ghpb.GetA
 	return &ghpb.GetAccessibleReposResponse{RepoUrls: urls}, nil
 }
 
+func (a *GitHubApp) CreateRepo(ctx context.Context, req *rppb.CreateRepoRequest) (*rppb.CreateRepoResponse, error) {
+	tu, err := a.env.GetUserDB().GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if tu.GithubToken == "" {
+		return nil, status.UnauthenticatedError("github account link is required")
+	}
+
+	// Pick the right client based on the request (organization or user).
+	githubClient, token, err := a.newAuthenticatedClient(ctx, tu.GithubToken)
+	if req.Organization != "" {
+		githubClient, token, err = a.newInstallationClient(ctx, req.InstallationId)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new repository on github
+	repo, _, err := githubClient.Repositories.Create(ctx, req.Organization, &github.Repository{
+		Name:        github.String(req.Name),
+		Description: github.String(req.Description),
+		Private:     github.Bool(req.Private),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// If we have a template, copy the template contents to the new repo
+	if req.Template != "" {
+		tmpDir := fmt.Sprintf("%s/template-repo/%d/%s/", scratchspace.TempDir(), req.InstallationId, req.Name)
+		err := cloneTemplate(tmpDir, token, req.Template, *repo.CloneURL)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Link new repository to enable workflows
+	wfs := a.env.GetWorkflowService()
+	if wfs == nil {
+		return nil, status.UnimplementedErrorf("no workflow service configured")
+	}
+	_, err = a.LinkGitHubRepo(ctx, &ghpb.LinkRepoRequest{
+		RequestContext: req.RequestContext,
+		RepoUrl:        *repo.HTMLURL,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &rppb.CreateRepoResponse{
+		RepoUrl: *repo.HTMLURL,
+	}, nil
+}
+
+func cloneTemplate(tmpDir, token, srcURL, destURL string) error {
+	// Make a temporary directory for the template
+	err := os.MkdirAll(tmpDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	dir, err := os.MkdirTemp(tmpDir, "repo")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	// Clone the template into the directory
+	auth := &githttp.BasicAuth{
+		Username: "github",
+		Password: token,
+	}
+	gitRepo, err := git.PlainClone(dir, false, &git.CloneOptions{
+		URL:  srcURL,
+		Auth: auth,
+	})
+	if err != nil {
+		return err
+	}
```

Comment:
should we maybe do this clone as a remote action (or add a TODO to do that soon?) Seems like these clones would cause bursts of CPU/disk usage (due to decompressing / unpacking git objects) and increase our cache tail latency.

The action could be a simple shell script (like `Arguments: []string{"bash", "-c", "git clone ... && git push ..."`), setting the token via env-var-overrides header like we do for workflows.

Reply:
Added a TODO. My biggest concern with doing that would be latency (I don't want a new user to sit and wait in a queue in this onboarding flow) and I think these templates should be relatively small at least to start. Maybe we could add some sort of task priority system in the future like Interactive / Non-interactive that would give these tasks priority.